### PR TITLE
send smtp email through a proxy

### PR DESCRIPTION
**Pull request by qdongxu**
This PR addresses the issue  #27348.  
The notification mailing will leverage the environment variable, `ALL_PROXY` if it is exported.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
"modules/proxy" should be improved, instead of introducing a totally new mechanism.

**Comment by lunny** [comment]
Looks like it's not finished.

**Comment by silverwind** [comment]
1. Uppercase proxy vars are discouraged as per [this](https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/), lowercase should be preferred.
2. I'm worried about `all_proxy` taking over `https_proxy` once https://github.com/golang/go/issues/31813 is implemented. Shouldn't it be called `smtp_proxy`?

**Comment by qdongxu** [comment]
> Looks like it's not finished.

It's finished. the `golang.org/x/net/proxy` package support ALL_PROXY/all_proxy environment variable out of the box.

**Comment by qdongxu** [comment]
according to  @wxiaoguang and @silverwind's comments, the strategy is to configure proxies in app.ini prior to the shell environment variable in order to behave clearly.   the `modules/proxy` handles http_proxy only for git operations, but stmp can only leverage socks proxy.  and it's a separate and irrelative decision to proxy git or mail. I'd like to add a `[smtp_proxy]` section, analogizing the `[proxy]` section.

**Comment by silverwind** [comment]
A specific ini option sound much better then relying on `all_proxy`, which seems a too dangerous variable beause of it's possible influence on other protocols like http.

Apparently there is already a known bug in golang related to the variable: https://github.com/golang/go/issues/16715

**Comment by wxiaoguang** [comment]
> but stmp can only leverage socks proxy.

That's not true. HTTP CONNECT proxy can also be used for any TCP protocol.

> and it's a separate and irrelative decision to proxy git or mail.

From a system design view, the Gitea should have the consistent proxy behavior.

> I'd like to add a `[smtp_proxy]` section, analogizing the `[proxy]` section.

Although at the moment I won't say "block" for it, but I really don't like this strange (inconsistent) behavior.

**Comment by wxiaoguang** [comment]
Again, such design doesn't look good to me. Could you explain why the `[proxy]` couldn't be reused? From my understanding, it should work (by making some improvements) with SMTP.

image

**Comment by qdongxu** [comment]
1. Add [stmp_proxy]: enabled property. proxy will only take effect when it is set to true.  either the  [stmp_proxy]:proxy_url or the all_prxoy environment will be ignore when  [stmp_proxy]: enabled  is false. 

2. about @silverwind 's concern, all_proxy environment variables may impact other protocols, this will not happen as I checked the going library, package golang.org/x/net/http/httpproxy, which only reads http_proxy/https_proxy environment variables, while package golang.org/x/net@v0.14.0/proxy, reads all_proxy only.  there would be no conflict. both [proxy] and [stmp_proxy] follow the same rule:
    1. if enabled is false, do not setup any proxy.
    2.  if enabled is true, read properties from the configuration file prior to environment variables.
    3. if enabled is true and file properties is empty,  read from the environment. 

4. about @wxiaoguang's comment http CONNECT can proxy for any TCP traffic, that's true after the connection is established, but the **CONNECT** here is a [http method CONNECT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) instead of a TCP connect invocation. for the moment I evaluated golang proxy( socks5 proxy only ), golang httpproxy ( bind to http.Client) and https://github.com/elazarl/goproxy , I did not find a way to use httpproxy for a non-http traffic. maybe I am wrong but an example would be appreciated. 

5. about @wxiaoguang's comment,  the Gitea should have consistent proxy behavior. I agree on it. but according to (4),   the most consistent is to make the properties in [proxy] and [smtp] section look similar and behave similarly. 

6. In [proxy] there's a ProxyHosts property, if host in ProxyHosts use proxyURL from config file , otherwise use http_proxy from the environment variable( and implies using the no_proxy environment variable ). This looks confusing, but this can be addressed in another issue.

**Comment by wxiaoguang** [comment]
I am pretty sure implementing the HTTP CONNECT proxy support is the right approach, and it's not difficult, like this: https://gist.github.com/jim3ma/3750675f141669ac4702bc9deaf31c6b .

HTTP CONNECT proxy is exactly (almost) the same as SOCSK5H proxy.

**Comment by qdongxu** [comment]
modules/proxy/httpsdialer.go was owned to https://gist.github.com/jim3ma/3750675f141669ac4702bc9deaf31c6b

updated to reuse [proxy]. and add a property `SMTP_PROXY_ENABLED` default as false, for backward compatibility.

**Review by wxiaoguang — APPROVED** [review]
The Gitea's proxy system has many legacy problems.

I vote my approval here and will do some improvements in following PRs.

File: modules/proxy/proxy_test.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package proxy
```
Comment (lunny) [inline modules/proxy/proxy_test.go:4]: Copy right header is necessary for linter.

Comment (qdongxu) [reply modules/proxy/proxy_test.go:4]: updated.

File: modules/proxy/proxy_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
```
Comment (techknowlogick) [inline modules/proxy/proxy_test.go]:
```suggestion
// Copyright 2023 The Gitea Authors. All rights reserved.
```

File: custom/conf/app.example.ini [hunk 3]
```diff
@@ -2559,6 +2559,9 @@ LEVEL = Info
 ;PROXY_URL =
 ;; Comma separated list of host names requiring proxy. Glob patterns (*) are accepted; use ** to match all hosts.
 ;PROXY_HOSTS =
+;; Enable the smtp proxy through when the proxy type is https://, socks://, all notifications through smtp emails
+;; will be affected. SMTP_PROXY_ENABLED takes effect with the precondition PROXY_ENABLED = true.
+;SMTP_PROXY_ENABLED = false
```
Comment (silverwind) [inline custom/conf/app.example.ini:2572]: Also sync this into docs/content/administration/config-cheat-sheet.en-us.md

**Comment by silverwind** [comment]
> about @silverwind 's concern, all_proxy environment variables may impact other protocols, this will not happen as I checked the going library, package golang.org/x/net/http/httpproxy, which only reads http_proxy/https_proxy environment variables, while package golang.org/x/net@v0.14.0/proxy, reads all_proxy only. there would be no conflict.

If there is no interference (e.g. `https_proxy` and `http_proxy` must take precedence over `all_proxy), I'm happy to support it as well.

**Comment by qdongxu** [comment]
Now the solution is just to reuse the modules/proxy.go to retrieve a proxyURL ( or "" ),  modules/proxy.go only reads http_proxy/https_proxy, not reads all_proxy.


> > about @silverwind 's concern, all_proxy environment variables may impact other protocols, this will not happen as I checked the going library, package golang.org/x/net/http/httpproxy, which only reads http_proxy/https_proxy environment variables, while package golang.org/x/net@v0.14.0/proxy, reads all_proxy only. there would be no conflict.
> 
> If there is no interference (e.g. `https_proxy` and `http_proxy` must take precedence over `all_proxy`), I'm happy to support it as well. Maybe we can leverage golang's environment detection so we don't have to support it in first-party code?

File: custom/conf/app.example.ini (hunk 3 above)
Comment (qdongxu) [reply custom/conf/app.example.ini:2572]: updated.

Comment (silverwind) [reply custom/conf/app.example.ini:2572]: thanks

**Comment by silverwind** [comment]
> > > about @silverwind 's concern, all_proxy environment variables may impact other protocols, this will not happen as I checked the going library, package golang.org/x/net/http/httpproxy, which only reads http_proxy/https_proxy environment variables, while package golang.org/x/net@v0.14.0/proxy, reads all_proxy only. there would be no conflict.
> > 
> > 
> > If there is no interference (e.g. `https_proxy` and `http_proxy` must take precedence over `all_proxy`), I'm happy to support it as well. Maybe we can leverage golang's environment detection so we don't have to support it in first-party code?
> 
> Now the solution totally reuses the modules/proxy.go to retrieve a proxyURL ( or "" for direct ), modules/proxy.go only reads http_proxy/https_proxy, not reads all_proxy.

Right, not a requirement for me.

File: modules/proxy/smtpproxy.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package proxy
+
+import (
+	"bufio"
+	"code.gitea.io/gitea/modules/setting"
```
Comment (lunny) [inline modules/proxy/smtpproxy.go]: import sequence is not right.

Comment (silverwind) [reply modules/proxy/smtpproxy.go]: Needs a `make fmt`, but there's more linter errors too.

Comment (qdongxu) [reply modules/proxy/smtpproxy.go]: updated. thanks.

File: custom/conf/app.example.ini [hunk 5]
```diff
@@ -2560,6 +2560,9 @@ LEVEL = Info
 ;PROXY_URL =
 ;; Comma separated list of host names requiring proxy. Glob patterns (*) are accepted; use ** to match all hosts.
 ;PROXY_HOSTS =
+;; Enable the smtp proxy through when the proxy type is https://, socks://, all notifications through smtp emails
+;; will be affected. SMTP_PROXY_ENABLED takes effect with the precondition PROXY_ENABLED = true.
```
Comment (silverwind) [inline custom/conf/app.example.ini]:
```suggestion
;; Enable proxying SMTP when PROXY_URL starts with https:// or socks://
```

Much simpler and says the same basically.

File: docs/content/administration/config-cheat-sheet.en-us.md [hunk 6]
```diff
@@ -1373,13 +1373,15 @@ is `data/repo-archive` and the default of `MINIO_BASE_PATH` is `repo-archive/`.
 - `PROXY_ENABLED`: **false**: Enable the proxy if true, all requests to external via HTTP will be affected, if false, no proxy will be used even environment http_proxy/https_proxy
 - `PROXY_URL`: **_empty_**: Proxy server URL, support http://, https//, socks://, blank will follow environment http_proxy/https_proxy
 - `PROXY_HOSTS`: **_empty_**: Comma separated list of host names requiring proxy. Glob patterns (*) are accepted; use ** to match all hosts.
+- `SMTP_PROXY_ENABLED`: **false**: Enable the smtp proxy when it's **true** and the proxy type is https://, socks://. All notifications through smtp emails will be affected. SMTP_PROXY_ENABLED takes effect with the precondition PROXY_ENABLED = **true**.
```
Comment (silverwind) [inline docs/content/administration/config-cheat-sheet.en-us.md]:
```suggestion
- `SMTP_PROXY_ENABLED`: **false**: Enable proxying SMTP when PROXY_URL starts with `https://` or `socks://`.
```

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
Still lint failures.

File: modules/proxy/httpsdialer.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	// HACK. http.ReadRequest also does this.
+	reqURL, err := url.Parse("http://" + addr)
+	if err != nil {
+		return nil, err
+	}
+	reqURL.Scheme = ""
+
+	req, err := http.NewRequest("CONNECT", reqURL.String(), nil)
```
Comment (silverwind) [inline modules/proxy/httpsdialer.go:60]: Seems we could use  [`httplib.NewRequest`](https://github.com/go-gitea/gitea/blob/6c501b149809a7015e58d9627e84bfd593702f95/modules/httplib/request.go#L36) instead, which also brings the benefit of setting a common user-agent.

Comment (qdongxu) [reply modules/proxy/httpsdialer.go:60]: Sorry for being late. As I check the `httplib.NewRequest`,  It aims to wrap a normal `HTTP request` and it uses the HTTP proxy environment variable internally :

```go
		TLSClientConfig: r.setting.TLSClientConfig,
		Proxy:           http.ProxyFromEnvironment,

```

while herein the `proxy/httpsdialer.go` just make a `CONNECT` to the proxy per se.

**Comment by qdongxu** [comment]
An alternative solution works : https://github.com/ankraft/smtpproxy
the issue and pr was closed thereby. thanks everyone reviewing the solution and code anyway.

**Pull request closed without merging**